plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/domains/CastingDomain.kt

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright 2010-2023 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.formver.domains
+
+import org.jetbrains.kotlin.formver.embeddings.TypeEmbedding
+import org.jetbrains.kotlin.formver.viper.ast.*
+
+/**
+ * Casting domain to cast any type to any other type
+ *
+ * Viper Domain:
+ * ```
+ * domain Casting[A, B]  {
+ *
+ *   function cast(a: A, newType: Type): B
+ *
+ *   axiom null_cast {
+ *     (forall newType: Type :: { (cast((null_val(): Nullable[A]), NullableType(newType)): Nullable[B]) }
+ *       (cast((null_val(): Nullable[A]), NullableType(newType)): Nullable[B]) == (null_val(): Nullable[B]))
+ *   }
+ *
+ *   axiom type_of_cast {
+ *     (forall a: A, newType: Type :: { (typeOf((cast(a, newType): Nullable[B])): Type) }
```

Comment:
Why is there a `Nullable[B]` here?  In particular, since you assume that the cast always succeeds.  Is this only the axiom for the nullable case, or does casting always produce a nullable value?

Reply:
You're right this should just be `B`.